### DISCUSSION TITLE
Remote code editing via `distant`

### DISCUSSION THREAD

**Discussion by chipsenkbeil** [helix-editor/helix#10411] [Ideas]
Curious if there's been any thought to offering remote code editing support similar to https://code.visualstudio.com/docs/remote/remote-overview.

I've been working on a project for a couple of years called [distant](https://github.com/chipsenkbeil/distant) with a plugin ([distant.nvim](https://github.com/chipsenkbeil/distant.nvim)) geared to leverage it to introduce remote coding editing support to neovim.

Since helix is written in Rust, it could leverage directly the libraries that distant offers to enable remote editing - including remote LSPs and shells - through both `ssh` and the distant protocol.

**Comment by archseer** [comment]
Wouldn't using something like sshfs to mount the remote fs be sufficient?

**Comment by chipsenkbeil** [comment]
@archseer, here's some of the primary differences that come to mind at the moment:

1. **Filesystem**: sshfs is built using FUSE, meaning that you need support for it as a userland filesystem. Having a native integration (i.e. like distant.nvim for neovim) requires no special filesystem as the file contents are only reflected in buffers. Anything you read goes into a buffer and anything you write gets transmitted across the network to the remote file.
2. **Program locations**: having a direct integration supports running programs on the remote machine, colocated with your files. Language servers would run on the remote machine, for instance. When using sshfs to mount the remote file system on your local machine, you would then run programs locally and point them to the mounted files. 
    * i.e. with distant.nvim, you are leveraging the resources of a remote machine. This can be advantageous if you want to perform CPU or GPU intensive operations without heavily impacting your local machine.
    * With sshfs, you are leveraging the resources of a local machine for programs that you run. Depending on what you're doing, this can be much more expensive. If a program needs to access the filesystem frequently, this would transmit a lot of requests over the network, especially if working over a large set of files. With distant.nvim, all of the filesystem operations would happen directly on the remote machine and only the program's output would be transmitted over the network.
3. **Singular connection**: instead of needing to maintain a separate connection for a shell using `ssh` alongside your filesystem being mounted via `sshfs`, having an integration like distant lets both being transmitted over the same network connection. This can be handy when you need to perform extensive authentication prior to connecting and want to avoid managing multiple connections.

There are other technical differences, but the distinctions above are - at the moment - the differences that stand out. In some cases, using sshfs might be preferred, but for my use personally and at work this is the preferred method as language servers and other utilities cannot be run locally. :)

If you want to think of what this more closely relates to, then take a look at [VS Code Remote Development](https://code.visualstudio.com/docs/remote/remote-overview) or the formerly-active [Nuclide](https://github.com/facebookarchive/nuclide).

**Comment by nathaniel-brough** [comment]
I don't know anything about distant. But I regularly use vscode remote. As helix is more of a tui text editor wouldn't it make more sense to install it on your dev server and ssh into it? I know sshfs was mentioned but why not just plain old ssh? For reference the vscode-remote server is already bigger than the helix editor. I'm curious as to what use case you couldn't meet by doing this. 

I think the case for vscode-remote is a little different as it is a web-based GUI rather than a TUI. It might be worth considering in more detail when/if helix gets a GUI?

EDIT: I'm not a dev here, just a curios user.

**Comment by chipsenkbeil** [comment]
@silvergasp good questions! The remote code functionality is agnostic to whether you use it with a terminal editor or a graphical editor.

Now, you bring up a good point about an advantage of terminal editors is that you can install and run them on remote machines more easily than a graphical editor.

The reasons that I've found a solution like this preferred - and many of my coworkers as well - are that the latency of your network doesn't affect your typing experience and similarly disruptions of your network don't fully block you from editing.

I worked on a bus to and from work for many years before 2020 and the network reliability was poor, causing tools like ssh to yield a broken pipe and mosh to freeze frequently waiting for the connection to re-establish. In contrast, VS Code's remote development feature let me continue working on a file without having my cursor and text freeze as I went in-and-out of networks.

For some of my colleagues, they live and work in locations where networks are poor, meaning that they cannot maintain a reliable ssh connection and mosh causes their editing experience to be nearly unusable as it frequently would cut out for ~10-15 seconds.

**Comment by chipsenkbeil** [comment]
Oh, and while this may not be as relevant to Helix, another reason is to avoid needing to copy your configuration files over to another machine. With ssh, to get the same experience with your editor, you have to copy your configuration files and plugins over to the other machine. In contrast, VS Code's remote editor and other offerings like distant let you avoid that and maintain your configurations on your local machine, only outsourcing programs and source code to the remote machine.

**Comment by nathaniel-brough** [comment]
Oh yeah, that seems reasonable. I'd definitely be keen on an improvement in latency/consistency!

**Comment by kirawi** [comment]
Tentatively marking as a plugin unless otherwise changed.

**Comment by jfaz1** [comment]
> Wouldn't using something like sshfs to mount the remote fs be sufficient?

I use SSHFS at work with Helix and it's almost enough to make me contemplate going back to VS Code (for work) 😆 

The connection is stable, the main issue is I can't use the file picker or global search because there are so many files on the remote server it hangs Helix for minutes at a time, sometimes even crashing it. Even just typing `:o <space>` out of muscle memory makes me need to restart the editor. It's a big repo (not as big as you'd think) but even small ones are plagued by this problem. I'd definitely appreciate a solution for this!

**Comment by dead10ck** [comment]
>Even just typing :o <space> out of muscle memory makes me need to restart the editor. It's a big repo (not as big as you'd think) but even small ones are plagued by this problem.

Not to hand wave the issue, because there is room for improvement there, but this sounds unusual. Are you sure there isn't a problem with IO on your server?

**Comment by jfaz1** [comment]
> Not to hand wave the issue, because there is room for improvement there, but this sounds unusual. Are you sure there isn't a problem with IO on your server?

Yeah, we route our network through a VPN which isn't the fastest, which will definitely slow things down. But even then I added an exception out of curiosity and while it was noticeably faster, was still unusable in the grand scheme of things. For context, opening the file picker in the fully cloned local repository takes around a second. So it's definitely a large repo.  But I tried some more manageably-sized repos and it was still pretty sluggish.

**Comment by taylorpool** [comment]
Hi all, just curious if there have been any updates on this issue. I love Helix and am very interested in using this feature.

**Comment by chipsenkbeil** [comment]
> Hi all, just curious if there have been any updates on this issue. I love Helix and am very interested in using this feature.

I haven't revisited yet. Distant is maturing more and more over time. One of the last major features was persistent connections, meaning that if you lose your connection with the remote machine, it will be able to reconnect later without any extra work or reauthentication from the user's side.

I'd be happy to work with someone on the helix side to integrate this. Either by just supporting the client-component of the library, or even making it where having helix on both machines would let you set up helix as a server for other helix instances to connect to.

Either way, going to let the folks from the helix side reach out if there's interest and availability. :)

**Comment by luigifcruz** [comment]
I have another use case for this. I have to access servers that are on the other side of the planet. The latency between my machine and the server is often higher than 300 ms. This can be very annoying while trying to type on a terminal-based IDE, like Helix. The remote feature proposed by this issue has the potential to ameliorate these latency issues.

**Comment by beegan** [comment]
I also see a benefit to this. I have a remote environment running on Amazon Linux 2 that comes packaged with glibc 2.26

As a result, helix can't be run natively there so support for running it remotely from my actual device would be a super handy workaround.

**Comment by shane-tran-whitmire** [comment]
Another benefit would be for development on older systems that cannot support helix. I am using debian 9 (I know, don't tell me it sucks), and literally cannot install it. Would be nice if I could ssh into the system with helix.

**Comment by pascalkuthe** [comment]
helix runs just fine on older systems you just have to compile it yourself.

**Comment by shane-tran-whitmire** [comment]
> helix runs just fine on older systems you just have to compile it yourself.

I might be a bit of an idiot, but I blew up (boot loader went bye bye) a system trying to get a new enough version of c++ to compile helix onto that system. My resolution was just to give up trying to get helix to run on debian 9.

**Comment by pascalkuthe** [comment]
That shuldn't happen. iiirc all C++ compilers released since 2015 rhold support everything we need so I am surprised it doesn't work out to the box. It can be a bit of a pain to get newer compiler sunning  but I bootstrapped clang-15 on centos 7 so it's definitely possible. There are probably existing docker images for this. I would not recommend messing with a live system. 

I use [this docker image](https://github.com/pascalkuthe/ferris-ci/pkgs/container/ferris_ci_build_x86_64-unknown-linux-gnu) to build my other rust projects for ancient glibc versions. I am not sure if it has everything helix has but you could try that if you don't find anything else

**Comment by gyzerok** [comment]
Since potentially many of us are subscribed to the issue to see updates on the remote functionality, I kindly ask you to move unrelated discussions somewhere else to avoid unnecessary notifications for everybody.

**Comment by nyabinary** [comment]
[Zed is now open source.](https://zed.dev/blog/zed-is-now-open-source)
Maybe we can look and see what Zed did?

**Comment by TJTorola** [comment]
Zed doesn't have remote code editing either: https://github.com/zed-industries/zed/issues/5347

**Comment by nyabinary** [comment]
> Zed doesn't have remote code editing either: [zed-industries/zed#5347](https://github.com/zed-industries/zed/issues/5347)

Zed does have good collab editing, tho.

**Comment by kirawi** [comment]
Zed is GPL and based on the Matrix conversation we can't reference it at all without polluting Helix's licensing.

**Comment by lianghu** [comment]
[Lapce](https://github.com/lapce/lapce) has a remote code editing support although Lapce itself is not stable yet.

**Comment by tulilirockz** [comment]
If you guys want an alternative to sshfs that doesnt freeze up whenever you use the file picker, check this out, seems particularly good for this case! It is just a workaround for now, but still might help out https://github.com/kahing/catfs/

**Comment by Hckr1123** [comment]
Hi, I really liked Helix’s design, compared to other editors. 
Switching to Helix, I immediately benefited from:
1) Minimum necessary features, ready-to-use out-of-the-box without bloat.
2) No complicated and time-consuming configuration for basic functionality (unlike other editors).
3) Out-of-the-box theming and syntax highlighting.
4) Simplified vim-style motions.
5) Discoverability (of features), reducing the need to spend considerable time on learning and referring documentation 

**All these contribute to PRODUCTIVITY and EFFICIENCY**, which is what I feel majority of us are after when it comes to using vim-like editors, and Helix is enabling that. 

So far, I felt just one main limitation when using Helix --- the lack of remote code editing and execution. 
**I feel more than a plugin, this deserves to be a built-in feature as it can really maximize Helix’s utility.**

It would address some significant limitations (like described in previous comments):
1) Productivity killing latency issues associated with editing remote server files over `sshfs`
2) Difficulty setting up Helix on restricted HPC/institutional servers (no sudo/admin privileges) and on servers that use older Linux versions/dependencies for stability/security/compatibility reasons
3) Clipboard limitation when running Helix on remote server, where copying to system clipboard would only copy the text to the server's clipboard and not that of the local machine (sometimes servers may not even have a proper clipboard facility set up).
4) Need for a separate ssh connection to remote server for server-side code execution (when using `sshfs` to edit code on remote server). Remote server side code execution is useful because often in the industry and academia, servers are/have the better performing compute --- High-End CPUs, GPUs, FPGAs, Custom-Hardware, and it enables us to leverage the server resources without leaving the code editor. This limitation may not be a productivity killer but disrupts the out-of-the-box complete/ready-to-use/streamlined experience that Helix enables for programmers (which other editors do not).

I feel incorporating this feature would not only boost Helix’s adoption but also make it more appealing for efficiency-oriented developers compared to using full-blown IDEs like VSCode. I am not an expert, and I may have overlooked some development considerations but I believe this is a crucial feature compared to other potential features like having a GUI or having plugin ecosystem support.

Who else feels this is a necessary feature to have? Please chime in so that the devs can know!

**Comment by prassanna-ravishankar** [comment]
This is super useful. Remote language servers along with a terminal (I know this is against the plan) will remove my dependency on vscode altogether.
